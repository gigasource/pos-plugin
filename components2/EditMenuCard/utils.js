export function createEmptyProductLayout() {
  return {
    type: 'Article',
    text: '',
    color: '#ddd',
    product: createEmptyProduct()
  }
}

export function createEmptyProduct() {
  return {
    type: null,
    id: '',
    name: '',
    price: '',
    groupPrinter: null,
    groupPrinter2: null,
    isNoPrint: null,
    isItemNote: null,
    tax: null,
    tax2: null,
    category: null,
    isDivArticle: null
  }
}


export function createEmptyLayout(row, column) {
  return {
    top: row,
    left: column,
    name: ''
  }
}

export function createEmptyCategoryLayout() {
  return {
    rows: 10,
    columns: 6,
    color: '#FFF'
  }
}


export function isSameArea(area1, area2) {
  return area1.top === area2.top && area1.left === area2.left
}
