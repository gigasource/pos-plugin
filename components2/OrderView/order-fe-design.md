* [x] isMobile: pos-order -> share
* [_] currentOrder

* [x] isolate editMode -> another file (use PortalVue)
* [_] storeLocale

* [_] createOrder
* [_] addProduct
* [_] changeItemQuantity
* [_] lần đầu create trên frontend : how to know ??
* [_]

* [x] takeAway logic
* [_] date logic
* [x] course:
    * [x] logic: 0: takeAway, 1: default
* [x] modifiers
* [_] tiền trả lại cashback
* [_] separate

* [_] table truyền vào như thế nào ?
* voucher:
* [x] item.quantityModifier : 
* [x] order has change : +- quantity to sent
* [x] xử lý order đã có từ backend
* [x] mock_order 
* [x] logic: nếu như backend set quantity = 0 -> ko show ra, còn frontend -> vẫn show ra
* [_] split order
* [_] dialog singleton : how to open them
* [_] payment
* [_] saveOrderTable (sendToKitchen) refactor
* [_] takeAway -> table có tính chất takeAway : how to test vue route -> setup thử xem thế nào
* [_] posOrder2 test snapshot trước -> easy hơn
* [x] pos layout -> convert now
* [x] pos layout
  * [_] snapshot for ui
  * [x] singleton setting ui
  * [_] categories: test/mock-db
  * [_] products: test/mock-db
  * [_] actionMode: test/mock-db
* [_] split order
  * [_] test backend
* [_] test thu cho khi ghep vao no se nhu the nao ?

```jsx
//addVoucher
this.addProductToOrder({
  name: 'Voucher',
  price: +value,
  isVoucher: true
})
//redeemVoucher
this.addProductToOrder({
  name: 'Redeemed Voucher',
  price: -value,
  isVoucher: true
})
```


items -> nested items

Sidebar : header + treeview + footer
header: -> portal sidebarHeader
const sidebarHeader = function (children) {
  return <div>{children}</div>
}

treeview -> customize duoc 1 so vi tri

const footer  
