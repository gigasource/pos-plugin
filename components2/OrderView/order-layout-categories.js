import {computed} from "vue";
import {
  editable,
  fillMissingAreas,
  getAreaStyle,
  getGridTemplateFromNumber,
  orderLayout,
  selectedCategoryLayout,
  selectedProductLayout,
  view
} from "./pos-ui-shared";
import {category} from "./order-layout-setting-logic";
import {isSameArea} from "../../components/posOrder/util";


export function orderLayoutCategoriesFactory() {
  const categories = computed(() => {
    if (editable.value) {
      return fillMissingAreas(
        orderLayout.categories,
        orderLayout.columns,
        orderLayout.rows,
        true);
    }
    const a = orderLayout;
    return [...orderLayout.value.categories]
  });

  const categoryContainerStyle = computed(() => {
    if (category && category.type === 'vertical') {
      if (category.differentSize) {
        return {
          display: 'block',
          width: category.size
        }
      }
      return {
        display: 'grid',
        'grid-template-rows': `repeat(8, calc(12.5% - ${5 * 8 / 7}px))`,
        'grid-template-columns': '100%',
        'height': '100%',
        "grid-gap": '5px',
        width: category.size
      }
    }
    if (category && category.type === 'horizontal' && category.singleRow) {
      if (category.differentSize) {
        return {
          display: 'flex',
          maxWidth: '100%',
          overflow: 'auto',
          height: category.size
        }
      }
      return {
        display: 'grid',
        'grid-template-columns': `repeat(8, calc(12.5% - ${5 * 8 / 7}px))`,
        'grid-template-rows': '100%',
        "grid-gap": '5px',
        height: category.size
      }
    }
    return {
      display: 'grid',
      'grid-template-columns': getGridTemplateFromNumber(orderLayout.columns),
      'grid-template-rows': getGridTemplateFromNumber(orderLayout.rows),
      'grid-gap': '5px',
      height: category ? category.size : `${40 * orderLayout.rows}px`
    }
  })

  function getCategoryStyle(category) {
    const isCategorySelected = selectedCategoryLayout && isSameArea(selectedCategoryLayout, category);
    return {
      backgroundColor: category.color,
      color: '#000',
      border: `1px solid ${isCategorySelected ? '#1271FF' : 'transparent'}`,
      boxShadow: isCategorySelected ? '1px 0px 3px rgba(18, 113, 255, 0.36)' : 'none',
      ...category && category.differentSize && (category.type === 'horizontal' ? {marginRight: "5px"} : {marginBottom: "5px"}),
      ...category && category.fontSize && {fontSize: category.fontSize}
    }
  }

  function getCategoryAreaStyle(cateItem) {
    if (category && category.type === 'vertical') {
      return getAreaStyle(cateItem, true)
    } else {
      return getAreaStyle(cateItem)
    }
  }

  async function selectCategory(categoryLayout) {
    if (editable.value) {
      selectedCategoryLayout.value = categoryLayout;
      view.value = {name: 'CategoryEditor'}
    } else {
      selectedCategoryLayout.value = categoryLayout;
    }
    selectedProductLayout.value = null;
  }

  function getCategoryName(item) {
    if (item) return item.name
  }

  const renderCategories = () => (
    <div
      style={{padding: '4px', 'background-color': '#E0E0E0', ...!editable.value && {position: 'sticky', top: 0, 'z-index': 1}}}>
      <div style={categoryContainerStyle}>
        {
          categories.value.map(category => (
            <div class='pol__cate'
                 style={[getCategoryStyle(category), getCategoryAreaStyle(category)]}
                 onClick={() => selectCategory(category)}>
              {getCategoryName(category)}
            </div>
          ))
        }
      </div>
    </div>
  )

  return {
    renderCategories
  }
}
