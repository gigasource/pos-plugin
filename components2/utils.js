import {computed, getCurrentInstance, ref, watch, withScopeId} from "vue";
import {useI18n} from "vue-i18n";
import dayjs from "dayjs";
import {useRouter} from 'vue-router';
import cms from 'cms';

export const internalValueFactory = (props, {emit}) => {
  const rawInternalValue = ref(props.modelValue);

  watch(() => props.modelValue, () => rawInternalValue.value = props.modelValue, { lazy: true });

  const internalValue = computed({
    get: () => rawInternalValue.value,
    set: (value) => {
      rawInternalValue.value = value;
      emit('update:modelValue', rawInternalValue.value)
    }
  });

  return internalValue;
}

let inited = false;

export function genScopeId(render, currentInstance) {
  if (!inited) initUtils();
  if (render) {
    return withScopeId((currentInstance || getCurrentInstance()).type['__scopeId'])(render);
  }
  return {[(currentInstance || getCurrentInstance()).type['__scopeId']]: ''};
}

export function execGenScopeId(render, currentInstance) {
  return genScopeId(render, currentInstance)()
}

export const parseNumber = function (number) {
  const res = parseFloat(number)
  if (isNaN(res)) return 0
  else return res
}

export const VModel_number = (obj, field = 'value') => {
  return computed({
    get: () => obj[field],
    set: (val) => obj[field] = parseNumber(val)
  })
}

export const intervalLog = (v, duration = 100) => {
  setInterval(() => console.log(v), duration)
}

export const isSameId = function (obj, obj1) {
  return obj._id.toString() === obj1._id.toString()
}

//todo dateFormat

export const formatsFactory = () => {
  const {t} = useI18n()
  const dateFormat = t('dates.dateFormat');
  return {dateFormat}
}

export const dateFormat = computed(() => {
  const {t} = useI18n();
  return t('dates.dateFormat');
});

export const timeFormat = computed(() => {
  const {t} = useI18n();
  return t('dates.timeFormat');
});

export const datetimeFormat = computed(() => {
  const {t} = useI18n();
  return `${t('dates.dateFormat')} ${t('dates.timeFormat')}`;
});

export function formatDate(date) {
  if (!date || !dayjs(date).isValid()) return ''
  return dayjs(date).format(dateFormat.value);
}

export function formatDatetime(date) {
  if (!date || !dayjs(date).isValid()) return ''
  return dayjs(date).format(datetimeFormat.value);
}

export function initUtils() {
  const [] = [
    dateFormat.value,
    timeFormat.value,
    datetimeFormat.value,
    backFn.value
  ]
  inited = true;
}

export const dateStringComputedFn = date => computed({
  get() {
    return dayjs(date.value).format('YYYY-MM-DD')
  },
  set(val) {
    date.value = dayjs(val, 'YYYY-MM-DD').toDate()
  }
});

export const backFn = computed(() => {
  const router = useRouter()

  function back() {
    router.push({path: '/pos-dashboard'})
  }

  return back;
})

export async function socketEmit() {
  return await new Promise(resolve => cms.socket.emit(...arguments, resolve));
}
