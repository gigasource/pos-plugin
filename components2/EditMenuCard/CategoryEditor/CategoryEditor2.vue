<script>
import ColorSelector from '../../../components/common/ColorSelector';
import InputNumber from '../../../components/EditMenuCard/InputNumber';
import PosKeyboardFull from '../../../components/pos-shared-components/PosKeyboardFull';
import { useI18n } from 'vue-i18n'
import constants from '../EditMenuCardToolbar/constants';
import { ref } from 'vue'

import {
  productRows,
  productCols,
  categoryName,
  categoryColor,
  debouncedUpdateCategory,
  deleteCategory,
  setAction,
  canSwitch,
  canDelete
} from './category-editor-category'

import {
  categoryRows,
  categoryColumns,
  changeCategoryColumn,
  changeCategoryRow,
} from './category-editor-order-layout'
import { genScopeId } from '../../utils';

export default {
  name: 'CategoryEditor2',
  components: { PosKeyboardFull, InputNumber, ColorSelector },
  setup() {
    const { t } = useI18n()

    // table layout setting: TODO: Move to order-layout-editor
    let renderCategoryLayoutSetting = () => <>
      <div class="category-editor__label">{t('restaurant.menuEdit.categoriesNo')}</div>
      <div class="row-flex align-items-center justify-between">
        <div class="fw-700 fs-small mr-2">{t('restaurant.menuEdit.columns')}:</div>
        <input-number
            modelValue={categoryColumns.value} min={1} max={8} width={148}
            onUpdate:modelValue={changeCategoryColumn}/>
      </div>
      <div class="row-flex align-items-center justify-between mt-1">
        <div class="fw-700 fs-small mr-2">{t('restaurant.menuEdit.rows')}:</div>
        <input-number
            modelValue={categoryRows.value} min={1} max={3} width={148}
            onUpdate:modelValue={changeCategoryRow}/>
      </div>
    </>

    // name setting
    const showCategoryNameKbd = ref(false)
    let renderCateNameSetting = () => (<>
      <div class="category-editor__label">{t('ui.name')}</div>
      <g-text-field-bs
          border-color="#979797"
          modelValue={categoryName.value}
          onUpdate:modelValue={newName => debouncedUpdateCategory({ name: newName }, newName)}
          v-slots={{
            'append-inner': () => <g-icon style="cursor: pointer" onClick={() => showCategoryNameKbd.value = true}>icon-keyboard</g-icon>
          }}>
      </g-text-field-bs>
    </>)

    // color setting
    const colors = ['#FFFFFF', '#CE93D8', '#B2EBF2', '#C8E6C9', '#DCE775', '#FFF59D', '#FFCC80', '#FFAB91']
    let renderCateColorSetting = () => (<>
      <div class="category-editor__label">{t('ui.color')}</div>
      <color-selector
          modelValue={categoryColor.value}
          colors={colors}
          itemSize={25}
          onUpdate:modelValue={color => debouncedUpdateCategory({color: color})}/>
    </>)

    // product layout setting
    let renderRowSetting = () => <>
      <div class="category-editor__label">{t('restaurant.menuEdit.rowsNo')}</div>
      <input-number
          modelValue={productRows.value} min={4} max={10}
          width={148}
          onUpdate:modelValue={newRow => debouncedUpdateCategory({rows: newRow})}/>
    </>
    let renderColsSetting = () => <>
      <div class="category-editor__label">{t('restaurant.menuEdit.columnsNo')}</div>
      <input-number
          modelValue={productCols.value} min={3} max={6}
          width={148}
          onUpdate:modelValue={newCol => debouncedUpdateCategory({columns:newCol})}/>
    </>

    let renderCateProductLayoutSetting = () => <>
      { renderRowSetting() }
      { renderColsSetting() }
    </>

    // pop-up
    let renderPopUp = () => <>
      <dialog-text-filter
          label={t('restaurant.menuEdit.categoryName')}
          defaultValue={categoryName.value}
          v-model={showCategoryNameKbd.value}
          onSubmit={newName => debouncedUpdateCategory({ name: newName}, newName /*forceCreate*/)}/>
    </>

    //// render toolbar buttons
    // delete button
    const showDeleteConfirmDialog = ref(false)
    let renderDeleteCategoryToolbarButton = () => {
      return <>
        <g-btn-bs elevation="2" icon="icon-edit-menu-card-delete"
                  onClick={() => showDeleteConfirmDialog.value = true}
                  disabled={!canDelete.value}>{t('ui.delete')}</g-btn-bs>
        <dialog-confirm-delete v-model={showDeleteConfirmDialog.value} type=' this category' onSubmit={() => {
          deleteCategory();
          showDeleteConfirmDialog.value = false;
        }}></dialog-confirm-delete>
      </>
    }
    let renderToolbarButtons = () => <portal to={constants.portalLeftButtons}>
      {
        genScopeId(() => <>
          <g-btn-bs elevation="2" icon="icon-edit-menu-card-switch"
                    onClick={() => setAction('switch')}
                    disabled={!canSwitch.value}>{t('ui.switch')}</g-btn-bs>
          { renderDeleteCategoryToolbarButton() }
        </>)()
      }
    </portal>

    // entire render
    let renderCategoryEditor = genScopeId(() => <div class="category-editor">
      {renderCategoryLayoutSetting()}
      {renderCateNameSetting()}
      {renderCateColorSetting()}
      {renderCateProductLayoutSetting()}
      {renderPopUp()}
      {renderToolbarButtons()}
    </div>)

    return {
      showDeleteConfirmDialog,
      showCategoryNameKbd,
      renderCategoryEditor
    }
  },
  render() {
    return this.renderCategoryEditor()
  }
}
</script>

<style scoped lang="scss">
  .category-editor {
    padding-left: 20px;
    padding-right: 20px;
    height: 100%;
    overflow: auto;

    &__label {
      color: #000;
      margin-top: 16px;
      margin-bottom: 13px;
      font-weight: 700;
      font-size: 15px;
    }

    .bs-tf-wrapper {
      margin: 0;

      ::v-deep .bs-tf-input {
        color: #000;
      }
    }
  }

  .g-snack ::v-deep .g-snack-wrapper {
    min-width: auto;
  }

  @media screen and (max-width: 1023px) {
    .category-editor {
      padding-left: 8px;
      padding-right: 8px;

      &__label {
        font-size: 13px;
        margin-top: 8px;
        margin-bottom: 4px;
      }
    }
  }

  @import "../EditMenuCardToolbar/emc-button";
</style>
