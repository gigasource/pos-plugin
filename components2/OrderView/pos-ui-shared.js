import {computed, nextTick, reactive, ref, watchEffect} from 'vue';
import {
  createEmptyCategoryLayout,
  createEmptyLayout,
  createEmptyProductLayout,
  isSameArea
} from "../../components/posOrder/util";
import _ from "lodash";
import {addItem} from "./pos-logic";
import { addProduct, getCurrentOrder, prepareOrder } from './pos-logic-be';
import {$filters} from "../AppSharedStates";
import {useI18n} from "vue-i18n";
import { createEmptyProduct } from '../EditMenuCard/utils';
import cms from 'cms';
export const orderLayout = ref({ categories: [] });
export const selectedCategoryLayout = ref();
export const selectedProductLayout = ref();
export const selectedProduct = computed({
  get: () => {
    // if no product layout selected, then return null
    if (!selectedProductLayout.value)
      return null

    // otherwise, init new product object if current product layout is empty
    if (!selectedProductLayout.value.product)
      selectedProductLayout.value.product = createEmptyProduct()

    return selectedProductLayout.value.product
  },
  set: (value) => {
    selectedProductLayout.value.product = value
  }
})
export const selectedProductExisted = computed(() => {
  return !!(selectedProduct.value && selectedProduct.value._id)
})

// editor view
export const view = ref({});
/**
 * Update editor display
 * @param name CategoryEditor | ProductEditor | KeyboardEditor
 * @param mode basic, ingredient
 */
export const updateView = (name, mode) => {
  mode = mode || (view.value && view.value.mode) // keep mode persistent

  // fallback to basic mode
  // we may see some thing weird like { name: CategoryEditor, mode: 'basic' }
  // but it's ok
  if (!mode)
    mode = 'basic'

  view.value = { name, mode }
}
export const ProductEditModes = {
  basic: 'basic',
  ingredient: 'ingredient'
}
export function updateProductEditMode(newMode) {
  view.value.mode = newMode
}

export const editable = ref(false);
export const productDblClicked = ref(false);

export function updateOrderLayout(newLayout) {
  orderLayout.value = newLayout
}
export function updateSelectedCategoryLayout(newCategoryLayout) {
  selectedCategoryLayout.value = newCategoryLayout
}
export function updateSelectedProductLayout(newProductLayout) {
  selectedProductLayout.value = newProductLayout
}

watchEffect(() => {
  if (!orderLayout.value)
    return

  if (selectedCategoryLayout.value) {
    // update category layout to force re-render after product action executed
    const cateLayout = _.find(orderLayout.value.categories, c => isSameArea(selectedCategoryLayout.value, c))
    if (cateLayout)
      updateSelectedCategoryLayout(cateLayout)
  } else {
    // automatically select first category
    if (orderLayout.value.categories.length > 0) {
      // find tab-product at 0-0
      const topLeftCategory = _.find(orderLayout.value.categories, c => c.top === 0 && c.left === 0)
      if (topLeftCategory)
        updateSelectedCategoryLayout(topLeftCategory)
      else
        updateSelectedCategoryLayout(_.first(orderLayout.value.categories))

      if (editable.value && (!view.value || view.value.name !== 'CategoryEditor'))
        updateView('CategoryEditor')
    }
  }
})

// TODO: rename -> productLayouts
export const products = computed(() => {
  if (!selectedCategoryLayout.value)
    return

  if (editable.value) {
    return fillMissingAreas(
      selectedCategoryLayout.value.products,
      selectedCategoryLayout.value.columns,
      selectedCategoryLayout.value.rows);
  }
  // remove product layout which is not text but doesn't link to any product
  return  _.filter(selectedCategoryLayout.value.products, p => p.type === 'Text' || (p.type !== 'Text' && p.product))
})

//fixme: only for dev

prepareOrder();
const order = getCurrentOrder();
const once = _.once(() => {
  addProduct(order, products.value[0].product);
  addProduct(order, products.value[0].product);
  addProduct(order, products.value[1].product);
  addProduct(order, products.value[1].product);

  orderViewDialog.move = true;
})
watchEffect(() => {
  if (order.items.length === 0 && products.value && products.value.length > 0) {
    once();
  }
})

/*async function run() {
  await nextTick();
}
run();*/

export const highlightSelectedProduct = ref(false);

export function getAreaStyle(item, rotate) {
  if (!item.top && item.top !== 0) {
    return
  }
  if (rotate) {
    return {
      'grid-area': `${item.left + 1} / ${item.top + 1} / ${item.left + 2} / ${item.top + 2}`
    }
  } else {
    return {
      'grid-area': `${item.top + 1} / ${item.left + 1} / ${item.top + 2} / ${item.left + 2}`
    }
  }
}

export function fillMissingAreas(areas, columns, rows, isCategory) {
  // add extra info
  // areas: [ { top, left }, {...} ]
  // generating empty elements
  const allAreas = [];
  for (let row = 0; row < rows; row++) {
    for (let column = 0; column < columns; column++) {
      let empty = createEmptyLayout(row, column);
      if (isCategory) {
        if (selectedCategoryLayout && isSameArea(empty, selectedCategoryLayout))
          empty = selectedCategoryLayout;
        else
          empty = {...empty, ...createEmptyCategoryLayout()}
      } else {
        if (selectedProductLayout && isSameArea(empty, selectedProductLayout))
          empty = selectedProductLayout
        else
          empty = {...empty, ...createEmptyProductLayout()}
      }
      allAreas.push(empty)
    }
  }
  return removeOuterAreas(_.unionWith(areas, allAreas, (area1, area2) => isSameArea(area1, area2)), columns, rows)
}

function removeOuterAreas(areas, columns, rows) {
  return _.filter(areas, area => area.top < rows && area.left < columns)
}

export function getGridTemplateFromNumber(num) {
  return _.join(_.map(_.range(0, num), c => `calc(${100 / num}% - ${5 * (num - 1) / num}px)`), ' ')
}

export function getItemSubtext({course, takeAway, separate}) {
  if (separate) return
  if (takeAway) return 'Take-away'
  if (course && course > 1) return `Course: ${course}`
}

export function isItemDiscounted(item) {
  return item.originalPrice && item.originalPrice !== item.price
}

export const orderViewDialog = reactive({
  search: false,
  split: false,
  move: false,
  voucher: false,
  receipt: false
});

export function itemsRenderFactory() {
  const {t, locale} = useI18n();
  const itemsRender = (items, onClick) => items.map(item => (
    <div class="item" key={item._id.toString()}>
      <div class="item-detail" onClick={() => onClick(item)}>
        <div>
          <p class="item-detail__name" style={[item.sent && {opacity: 0.55}]}>
            {item.id && `${item.id}. `}{item.name}</p>
          <p>
            <span class={['item-detail__price', isItemDiscounted(item) && 'item-detail__discount']}>
              {t('common.currency', locale)}{$filters.formatCurrency(item.originalPrice)}
            </span>
            {isItemDiscounted(item) &&
            <span class="item-detail__price--new">
              {t('common.currency', locale)} {$filters.formatCurrency(item.price)}
            </span>
            }
            <span class={['item-detail__option', 'text-red-accent-2']}>
              {getItemSubtext(item)}
            </span>
          </p>
        </div>
        <div class="mr-2 fw-700 row-flex align-items-center"
             style="font-size: 18px">{item.quantity}</div>
      </div>
      {item.modifiers && <div>
        {item.modifiers.map(modifier => (
          <g-chip label small text-color="#616161">
            {modifier.name} |
            {t('common.currency', locale)} {$filters.formatCurrency(modifier.price)}
          </g-chip>
        ))}

      </div>}
    </div>
  ))
  return itemsRender;
}
