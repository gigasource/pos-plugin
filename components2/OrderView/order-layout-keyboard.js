import cms from 'cms';
import {computed, reactive, ref} from "vue";
import {editable, selectedCategoryLayout} from "./pos-ui-shared";
import _ from "lodash";

export const keyboardConfig = ref();

const dialog = reactive({
  value: false,
  position: {},
  search: false
})

export async function loadKeyboardConfig() {
  const setting = await cms.getModel('PosSetting').findOne()
  const _keyboardConfig = setting.keyboardConfig;
  if (_keyboardConfig && !_keyboardConfig.layout)
    _keyboardConfig.layout = []
  keyboardConfig.value = _keyboardConfig || {layout: []};
}

const showCalculator = computed(() => {
  if (!selectedCategoryLayout.value || !keyboardConfig.value)
    return false
  //out of boundary
  if (selectedCategoryLayout.value.rows <= keyboardConfig.value.top
    || selectedCategoryLayout.value.columns <= keyboardConfig.value.left)
    return false

  let show = keyboardConfig.value.active
  if (keyboardConfig.value.onlyShowInFirstPage) {
    const {top, left} = selectedCategoryLayout.value
    if (top !== 0 || left !== 0)
      show = false
  }

  return show
})

export async function updateKeyboardConfig(keyboardConfig) {
  await cms.getModel('PosSetting').findOneAndUpdate({}, {$set: {keyboardConfig}})
}

async function changeKeyboardExtension(val) {
  const config = Object.assign({}, keyboardConfig.value)
  _.set(config, ['layout', dialog.position.top - 1, 'rows', dialog.position.left - 1], val)
  await updateKeyboardConfig(config)
  keyboardConfig.value = config;
}

function openDialogEdit(position) {
  dialog.position = position
  dialog.value = true
}

export function orderLayoutKeyboardFactory(editable) {
  const renderKeyboard = () => {
    return (
      <>
        {showCalculator.value && <pos-order-keyboard2 keyboard-config={keyboardConfig}
                             mode={editable ? 'edit' : 'active'}
                             onEdit:keyboard={e => openDialogEdit(e)}
                             onOpenDialogSearch={() => dialog.search = true}/>}
        <dialog-text-filter v-model={dialog.value} onSubmit={e => changeKeyboardExtension(e)}/>
        <dialog-product-search-result v-model={dialog.search}/>
      </>
    )
  }

  return {
    renderKeyboard
  }
}
