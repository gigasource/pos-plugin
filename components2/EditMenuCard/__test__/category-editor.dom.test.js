/**
 * NOTE:
 *
 * This file contain test cases to test whether CategoryEditor render correct or not
 *
 * With valid category layout, it should render order layout section in the top to allow
 * the user to modify the number of row or column of category will be display in orderLayout.
 *
 * It display category information (name, color, the number of row, column of product in
 *    this cate which corresponding to rows=4 & cols=6)
 *
 * It also tell us about actions which we can do with selected category:
 *   - switch if it's real category layout
 *   - delete if category doesn't have any product
 *
 * -----------
 *
 * It's doesn't contain any test cases to test actions (delete, switch) or change category information
 * It's assume that dependencies component (portal, 'dialog-confirm-delete', 'g-grid-select',
 *    'color-selector','input-number','dialog-text-filter', ... are render correctly with correct input.
 */
import {
  wrapper,
  makeWrapper,
} from "../../test-utils";
const delay = require("delay");
import CategoryEditor2 from '../CategoryEditor/CategoryEditor2';
import {nextTick} from 'vue'
import { orderLayout, selectCategoryLayout } from '../../OrderView/pos-ui-shared';
import { createEmptyCategoryLayout, createEmptyLayout } from '../utils';
const ignoreComponents = ['Portal', 'PortalTarget', 'MountingPortal', 'portal', 'portal-target']

describe('category-editor', () => {
  async function setup(selectedCategory) {
    orderLayout.value = {
      rows: 1,
      columns: 3,
      type: 'default',
      categories: [
        selectedCategory
      ]
    }

    selectCategoryLayout({ top: 0, left: 0 })

    // wipe-out components (to make dependencies component un-resolvable) to make sure snapshot render prop for stub class
    CategoryEditor2.components = {}

    makeWrapper(CategoryEditor2, {
      shallow: false,
      global: {
        stubs: {
          portal: true,
          'dialog-confirm-delete': true,
          'g-grid-select': true,
          'color-selector': true,
          'input-number': true,
          'dialog-text-filter': true,
        },
        config: {
          isCustomElement: tag => _.includes(ignoreComponents, tag)
        }
      }
    }, false)
    await nextTick()
    await delay(50)
    expect(wrapper.html()).toMatchSnapshot()
  }

  it('should render category with 0 product, allow to switch & delete', async () => {
    // can delete category without any products
    await setup({ _id: '012345', name: 'Wine', color: '#FFF', rows: 4, columns: 6, top: 0, left: 0 })

    // can delete category with empty product - which created in editable mode as a placeholder
    await setup({ _id: '012345', name: 'Wine', color: '#FFF', rows: 4, columns: 6, top: 0, left: 0,
      products: [
        { product: { name: 'prod1', price: 10 }, top: 0, left: 0 },
        { product: { name: 'prod1', price: 20 }, top: 0, left: 1 },
      ]
    })
  })

  it('should render category which has products, allow to switch but not delete', async () => {
    await setup({ _id: '012345', name: 'Wine', color: '#FFF', rows: 4, columns: 6, top: 0, left: 0,
      products: [
        { product: { _id: 'prod1', name: 'prod1', price: 10 }, top: 0, left: 0 }, // real product
        { product: { name: '', price: 0 }, top: 0, left: 1 } // placeholder product
      ]
    })
  })

  it('should render empty category correctly', async () => {
    await setup({...createEmptyLayout(0, 0), ...createEmptyCategoryLayout()})
  })
})
