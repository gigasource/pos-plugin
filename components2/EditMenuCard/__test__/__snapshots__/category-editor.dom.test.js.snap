// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`category-editor should show correct 1`] = `
<div class="category-editor__label">
  restaurant.menuEdit.categoriesNo
</div>
<div class="row-flex align-items-center justify-between">
  <div class="fw-700 fs-small mr-2">
    restaurant.menuEdit.columns:
  </div>
  <input-number-stub modelvalue="4"
                     min="1"
                     max="8"
                     width="148"
  >
  </input-number-stub>
</div>
<div class="row-flex align-items-center justify-between mt-1">
  <div class="fw-700 fs-small mr-2">
    restaurant.menuEdit.rows:
  </div>
  <input-number-stub modelvalue="2"
                     min="1"
                     max="3"
                     width="148"
  >
  </input-number-stub>
</div>
<div class="category-editor__label">
  ui.name
</div>
<g-text-field-bs-stub border-color="#979797"
                      modelvalue="Wine"
>
</g-text-field-bs-stub>
<div class="category-editor__label">
  ui.color
</div>
<color-selector-stub modelvalue="#FFF"
                     colors="#FFFFFF,#CE93D8,#B2EBF2,#C8E6C9,#DCE775,#FFF59D,#FFCC80,#FFAB91"
                     itemsize="25"
>
</color-selector-stub>
<div class="category-editor__label">
  restaurant.menuEdit.rowsNo
</div>
<input-number-stub modelvalue="4"
                   min="4"
                   max="10"
                   width="148"
>
</input-number-stub>
<div class="category-editor__label">
  restaurant.menuEdit.columnsNo
</div>
<input-number-stub modelvalue="6"
                   min="3"
                   max="6"
                   width="148"
>
</input-number-stub>
<dialog-text-filter-stub label="restaurant.menuEdit.categoryName"
                         defaultvalue="Wine"
                         modelvalue="false"
>
</dialog-text-filter-stub>
<portal-stub to="portal-left-buttons">
  <g-btn-bs-stub elevation="2"
                 icon="icon-edit-menu-card-switch"
                 disabled="true"
  >
    ui.switch
  </g-btn-bs-stub>
  <g-btn-bs-stub elevation="2"
                 icon="icon-edit-menu-card-delete"
                 disabled="true"
  >
    ui.delete
  </g-btn-bs-stub>
  <dialog-confirm-delete-stub modelvalue="false"
                              type=" this category"
  >
  </dialog-confirm-delete-stub>
</portal-stub>
`;
